tempesta_fw/http.c

```diff
@@ -2241,21 +2354,46 @@ tfw_http_resp_fwd(TfwHttpResp *resp)
 	if (unlikely(list_empty(seq_queue))) {
 		BUG_ON(!list_empty(&req->msg.seq_list));
 		spin_unlock_bh(&cli_conn->seq_qlock);
-		TFW_DBG2("%s: The client was disconnected, drop resp and req: "
-			 "conn=[%p]\n",
-			 __func__, cli_conn);
+		TFW_DBG2("%s: The client was disconnected conn=[%p], drop "
+			 "resp=[%p] and req=[%p]\n",
+			 __func__, cli_conn, resp, req);
 		ss_close_sync(cli_conn->sk, true);
-		tfw_http_resp_pair_free(req);
+		/* Don't destroy streamed response until streaming ended. */
+		if (!tfw_http_msg_is_streamed((TfwHttpMsg *)resp)
+		    || (resp->flags & TFW_HTTP_F_STREAM_DONE))
+		{
+			tfw_http_resp_pair_free(req);
+		}
 		TFW_INC_STAT_BH(serv.msgs_otherr);
 		return;
 	}
 	BUG_ON(list_empty(&req->msg.seq_list));
 	resp->flags |= TFW_HTTP_F_RESP_READY;
+	/* Last response was sent in streaming mode, hold on now. */
+	if (cli_conn->msg_stream) {
+		spin_unlock_bh(&cli_conn->seq_qlock);
+		return;
+	}
 	/* Move consecutive requests with @req->resp to @ret_queue. */
 	list_for_each_entry(req, seq_queue, msg.seq_list) {
 		if (!req->resp || !(req->resp->flags & TFW_HTTP_F_RESP_READY))
 			break;
+		/*
+		 * Request is streamed, don't send reply until request is
+		 * fully received.
+		 */
+		if (tfw_http_msg_is_streamed((TfwHttpMsg *)req)
+		    && !tfw_http_msg_is_processed((TfwHttpMsg *)req))
+		{
+			break;
+		}
```

Comment:
If a backend server already respond us, why should we wait for the end of the request?